drivers/char/optimsoc-noc.c

```diff
@@ -0,0 +1,436 @@
+/**
+ * Copyright (c) 2012-2017 by the author(s)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * =================================================================
+ *
+ * Linux driver for OpTiMSoC network-on-chip adapter.
+ *
+ * Author(s):
+ *   Pedro H. Penna <pedrohenriquepenna@gmail.com>
+ *   Stefan Wallentowitz <stefan.wallentowitz@tum.de>
+ */
+
+#include <asm/io.h>
+#include <asm/uaccess.h>
+#include <linux/fs.h>
+#include <linux/interrupt.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+
+/**
+ * Maximum number of endpoints in a tile.
+ */
+#define NR_ENDPOINTS 2
+
+/*
+ * Device name.
+ */
+#define OPTIMSOC_NA_NAME "optimsoc-noc"
+
+/*
+ * NoC adapter IRQ number.
+ */
+#define OPTIMSOC_NA_IRQ 4
+
+/*
+ * Base hardware address for NoC adapter.
+ */
+#define OPTIMSOC_NA_BASE_HWADDR 0xe0000000
+
+/**
+ * Base virtual address for NoC adapter
+ */
+static uint32_t OPTIMSOC_NA_BASE_VADDR = 0;
+
+/*
+ * NoC adapter memory size.
+ */
+#define OPTIMSOC_NA_MEM_SIZE (8*1024*1024)
+
+/* 
+ * Address of memory maped registers.
+ */
+#define REG_NUM_CT (OPTIMSOC_NA_BASE_VADDR + 0x28)
+#define BASE       (OPTIMSOC_NA_BASE_VADDR + 0x100000)
+#define REG_NUMEP  BASE
+#define EP_BASE    BASE + 0x2000
+#define EP_OFFSET  0x2000
+#define REG_SEND   0x0
+#define REG_RECV   0x0
+#define REG_ENABLE 0x4
+/**
+ * Retrieves a word at a target address.
+ */
+#define REG32(x) (*((uint32_t *)(x)))
```

Comment:
are there no kernel internal functions/macros for that?

Reply:
Usually we do `ioread32be()`